### PR TITLE
Debug info preservation in redundancy-elimination pass

### LINKED ISSUE #3821: [OpenCL.DebugInfo.100] support debug info preservation in redundancy-elimination pass

### PR DESCRIPTION
**Pull request by jaebaek**
Fixes #3821

### CODE REVIEW

File: test/opt/redundancy_elimination_test.cpp [hunk 1]
```diff
@@ -272,6 +272,59 @@ TEST_F(RedundancyEliminationTest, KeepRedundantAddWithoutPhi) {
...
+         %22 = OpExtInst %void %1 DebugLocalVariable %6 %19 %16 0 0 %21 FlagIsLocal
+          %3 = OpFunction %void None %9
+         %23 = OpLabel
+         %24 = OpExtInst %void %1 DebugScope %21
+         %25 = OpVariable %_ptr_Function_float Function
+         %26 = OpLoad %float %25
+               OpLine %4 0 0
+; CHECK:      [[add:%\w+]] = OpFAdd %float [[value:%\w+]]
+; CHECK-NEXT: DebugValue {{%\w+}} [[add]]
+; CHECK-NEXT: OpFAdd %float [[add]] [[value]]
```
Comment (s-perron) [inline test/opt/redundancy_elimination_test.cpp:313]: It is not clear that the section OpFAdd corresponds to the one just before the return.  A comment might be useful.

File: source/opt/local_redundancy_elimination.cpp [hunk 2]
```diff
@@ -58,6 +65,22 @@ bool LocalRedundancyEliminationPass::EliminateRedundanciesInBB(
       context()->ReplaceAllUsesWith(inst->result_id(), candidate.first->second);
       context()->KillInst(inst);
       modified = true;
+    } else if (inst->GetOpenCL100DebugOpcode() ==
+                   OpenCLDebugInfo100DebugDeclare ||
+               inst->GetOpenCL100DebugOpcode() ==
+                   OpenCLDebugInfo100DebugValue) {
+      uint32_t var_id =
```
Comment (s-perron) [inline source/opt/local_redundancy_elimination.cpp:72]: There might be a better way to handle this.  Why is the result of `GetValueNumber` different for the two redundant DebugValue instructions?  I would expect that we should change value number to make them the same.

**Review by s-perron — CHANGES_REQUESTED** [review]
Could you look into value numbering to see if a change there could work?  That way anything that depends on value number will be able to benefit from this change.

File: source/opt/local_redundancy_elimination.cpp (hunk 2 above)
Comment (jaebaek) [reply source/opt/local_redundancy_elimination.cpp]: done

File: test/opt/redundancy_elimination_test.cpp (hunk 1 above)
Comment (jaebaek) [reply test/opt/redundancy_elimination_test.cpp:320]: done

**Review by jaebaek — COMMENTED** [review]
Thank you for your code review.
I updated `ValueNumberTable` to support OpenCL.DebugInfo.100 instructions.
PTAL!

**Review by s-perron — COMMENTED** [review]
Looks good, but add some value numbering tests as well to make sure that value numbering is doing the right thing: https://github.com/KhronosGroup/SPIRV-Tools/blob/master/test/opt/value_table_test.cpp

**Comment by jaebaek** [comment]
Added unit tests to ValueTableTest. PTAL!

**Review by s-perron — APPROVED** [review]

**Pull request merged**